### DISCUSSION TITLE
How to turn system monitor script into a app?

### DISCUSSION THREAD

**Discussion by HugoP** [h2oai/wave#2245] [Q&A] [answered]
Hi, I'm new to wave and started running some of the tutorials but I'm stuck with turning a script into a app.

The script in question is the [system_monitor](https://wave.h2o.ai/docs/tutorial-monitor) example.

How can I turn this into a non blocking app?

Any help will be appreciated.

**Comment by mturoci** [comment]
cc @marek-mihok

**Comment by marek-mihok** [comment, marked as answer]
The answer lies in unblocking `psutil.cpu_percent()` by removing its `interval=1` parameter and using asynchronous `q.sleep` instead of synchronous `time.sleep`. You can see it's non-blocking on this counter app:

https://github.com/h2oai/wave/assets/23740173/573ccd10-d95a-4ae9-b0f2-b60ad8ed451c


```py
import psutil

from h2o_wave import main, app, Q, ui, data

async def monitor_cpu(q: Q):
    while True:
        await q.sleep(1)
        cpu_usage = psutil.cpu_percent()
        q.page['monitor'].value = cpu_usage
        q.page['monitor'].data.usage = cpu_usage
        q.page['monitor'].plot_data[-1] = [q.client.tick, cpu_usage]
        q.client.tick += 1    
        await q.page.save()

@app('/monitor')
async def serve(q: Q):
    if not q.client.initialized:
        q.client.tick = 0
        q.client.count = 0
        q.page['monitor'] = ui.small_series_stat_card(
            box='1 1 1 1',
            title='CPU',
            value='={{usage}}%',
            data=dict(usage=0.0),
            plot_data=data('tick usage', -15),
            plot_category='tick',
            plot_value='usage',
            plot_zero_value=0,
            plot_color='$red',
        )
        q.page['counter'] = ui.form_card(box='2 1 2 2', items=[
            ui.text(name='text', content=f'Current count: {q.client.count}'),
            ui.button(name='increment', label='Increment'),
        ])
        q.client.initialized = True 
        await monitor_cpu(q)

    if q.args.increment:
        q.client.count += 1
        q.page['counter'].text.content = f'Current count: {q.client.count}'
    
    await q.page.save()
```

If you struggle with unblocking your apps in the future [this article](https://medium.com/@unusualcode/background-jobs-in-wave-or-how-not-to-kill-your-ui-ae1fed95693a) by @mturoci may be helpful.

**Comment by HugoP** [comment]
That's great, thanks.

**Reply by HugoP** [reply]
Am I correct that all wave app's need to be run with --no-reload to avoid getting stuck at Waiting for background tasks to complete?

```
INFO:     Shutting down
INFO:     Waiting for background tasks to complete. (CTRL+C to force quit)
```

**Reply by mturoci** [reply]
`--no-reload` means it doesn't watch for file changes to restart the server. If you do not want to wait for background tasks to complete, check uvicorn docs to see whether it's possible.
